src/atc/sql/schema.py

```diff
@@ -1,16 +1,315 @@
-from deprecated import deprecated
+import dataclasses
+import re
+import string
+
+from more_itertools import peekable
 from pyspark.sql import types as t
 
-from atc.spark import Spark
 
+class SchemaExtractionError(Exception):
+    pass
+
+
+__debug = False
+
+
+def log(*args, **kwargs):
+    if __debug:
+        print(*args, **kwargs)
+
+
+def get_schema(sql: str) -> t.StructType:
+    # remove witespace and parenthesis
+    sql = sql.strip().strip("()").strip()
+
+    # remove block comments
+    sql = re.sub(r"/\*.*?\*/", "", sql, flags=re.MULTILINE)
+
+    # remove line comments
+    sql = re.sub("--.*", "", sql, flags=re.MULTILINE)
+
+    # simplify all whitespace
+    sql = re.sub(
+        r"\s+", " ", sql, flags=re.MULTILINE
+    )  # make all whitespace a single space
+
+    for c in [",", ":", "<", ">", r"\(", r"\)"]:
+        # remove whitespace before and after spacial character
+        sql = re.sub(rf"\s*{c}\s*", c.strip("\\"), sql, flags=re.MULTILINE)
+
+    # we can go no further with regex
+
+    it = peekable(sql)
+    struct = t.StructType([])
+    try:
+        while True:
+            name = _get_identifier(it)
+            _peek_skip_space_or_colon(it)
+            data_type = _get_data_type(it)
+            m = _get_metadata(it)
+            metadata = {"comment": m.comment} if m.comment else None
+            struct.add(
+                field=name, data_type=data_type, nullable=m.nullable, metadata=metadata
+            )
+            if next(it) == ",":
+                continue
+            else:
+                raise SchemaExtractionError("Malformed SQL")
+    except StopIteration:
+        pass
+    return struct
+
+
+@dataclasses.dataclass
+class _Metadata:
+    comment: str = None
+    nullable: bool = True
+
+
+def _get_metadata(it) -> _Metadata:
+    m = _Metadata()
+    while it:
+        if it.peek() == ",":
+            break  # no more metadata
+
+        if it.peek() != " ":
+            raise SchemaExtractionError("Malformed SQL")
+        next(it)  # consume space
+
+        token = _get_ascii_token(it).upper()
+        if token == "NOT":
+            for c in " NULL":
+                try:
+                    if next(it).upper() != c:
+                        raise SchemaExtractionError("Malformed SQL")
+                except StopIteration:
```

Comment:
if we could give a hint on what is wrong with the malformatting- should we print it? It would make it easier for developers to debug. This comment also applies for the other usages of `raise SchemaExtractionError("Malformed SQL")`

Reply:
like the goo hint in line 170

Reply:
yes, but this is too challenging right now. it needs some upstream support for token positions. I will not wait for that right now.